[Refactor] 长期记忆实现重构，新增三种长期记忆引擎，优化提取 Prompt 和性能

Hoshino-Yumetsuki opened the pull request.

Hoshino-Yumetsuki committed:
refactor: lint

Hoshino-Yumetsuki committed:
feat(long-memory): 长期记忆增强，使用文本指纹提升召回质量

Hoshino-Yumetsuki committed:
feat(long-memory): 破坏性改动，切换长期记忆实现

dingyi222666 commented:
等待向量数据库重构完后继续处理

Hoshino-Yumetsuki committed:
refactor(long-memory): 增强鲁棒性

Hoshino-Yumetsuki committed:
style: lint

Hoshino-Yumetsuki committed:
refactor(long-memory): 优先使用 editDocument 更新访问统计

Hoshino-Yumetsuki committed:
Merge branch 'v1-dev' into v1-dev

Hoshino-Yumetsuki committed:
refactor(long-memory): 解决接口变动

Hoshino-Yumetsuki committed:
fix(long-memory): 修复类型不匹配

Hoshino-Yumetsuki committed:
refactor(long-memory): 解决 lint 错误

Hoshino-Yumetsuki committed:
Merge branch 'ChatLunaLab:v1-dev' into v1-dev

dingyi222666 committed:
feat(long-memory): add HippoRAG integration and remove legacy tools

- Add new HippoRAG implementation in src/rag/ directory
- Remove deprecated HippoRAG inspection tools from tool.ts
- Export createVectorStoreRetriever and TextTokenizer utilities
- Add memory documentation files (memory.md, memory.pdf)

This refactors the memory system to use a cleaner HippoRAG implementation
while removing unused administrative tools and exposing needed utilities.

dingyi222666 committed:
feat(long-memory): refactor memory layer architecture and improve configuration

- Restructure memory layers with unified layer architecture supporting Global, Preset, and User layers
- Replace legacy RAG implementation with modular layer engines (Basic, HippoRAG, Emgas)
- Remove deprecated PRESET_USER layer type in favor of separate Preset and User layers
- Add new configuration schema for layer engines with experimental HippoRAG and Emgas support
- Implement layer initialization service with proper conversation-scoped management
- Move HippoRAG IE and KG utilities to dedicated layer directories
- Add Emgas layer implementation with episodic memory graph and activation spreading
- Update all plugins to use new initMemoryLayers API instead of createMemoryLayers
- Add js-yaml dependency for YAML parsing in graph extraction
- Enhance database docstore with ID filtering support for improved retrieval

This refactor provides a more flexible and extensible architecture for long-term memory
management while maintaining backward compatibility for existing memory operations.

dingyi222666 commented on packages/long-memory/memory.md:
```diff
@@ -0,0 +1,161 @@
+  private nodes: Map<string, MemoryNode>;
+  private edges: Map<string, Map<string, MemoryEdge>>; // 邻接表示法
+
+  // 用于增量计算PMI的全局计数器
+  private conceptCounts: Map<string, number>;
+  private pairCounts: Map<string, number>; // key为 "concept1|concept2"
+  private totalObservations: number;
+
+  constructor() {
+    this.nodes = new Map();
+    this.edges = new Map();
+    this.conceptCounts = new Map();
+    this.pairCounts = new Map();
+    this.totalObservations = 0;
+  }
+
+  // 公共API方法
+  public incrementalUpdate(textChunk: string): void {
+    // 实现第3.3节中的增量更新逻辑
+  }
+
+  public retrieveContext(query: string, options: SpreadingActivationOptions): string {
+    // 实现第4章中的激活扩散检索逻辑
+    // 返回的是需要被提取的原始文本片段的ID数组
+  }
+
+  public applyDecay(lambda: number): void {
+    // 实现第5.1节中的指数衰减逻辑
+  }
+
+  public pruneGraph(threshold: number): void {
+    // 实现第5.2节中的图谱剪枝逻辑
+  }
+
+  //... 其他私有辅助方法，例如：
+  // private extractAndNormalizeConcepts(text: string): string {... }
+  // private calculatePPMI(concept1: string, concept2: string): number {... }
+  // private boostNodeActivation(nodeId: string): void {... }
+}
+6.2. 算法实现伪代码以下是MemoryGraph类中两个核心方法的详细伪代码，其逻辑清晰，可直接翻译为TypeScript代码。retrieveContext (包含激活扩散)代码段function retrieveContext(query, options):
```
无需理会

dingyi222666 commented on packages/long-memory/src/rag/lazer.ts:
```diff
@@ -0,0 +1,233 @@
+        )
+
+        return relevantDocs.map(documentToEnhancedMemory)
+    }
```
finish

dingyi222666 committed:
feat(long-memory): add EMGAS configuration parameters and improve layer architecture

- Add comprehensive EMGAS configuration schema with decay rate, pruning threshold, firing threshold, propagation decay, max iterations, and top N parameters
- Include proper min/max validation and default values for all EMGAS settings
- Add complete English and Chinese internationalization for EMGAS parameters
- Refactor EMGAS and HippoRAG layers to use configuration values instead of hardcoded constants
- Improve model reference handling in extraction modules for better type safety
- Fix vector store method signatures and database query optimization
- Add proper memory ID generation for enhanced memory objects
- Reorganize configuration structure for better maintainability
- Update chat middleware to use configurable query rewrite setting

This enhancement provides fine-grained control over EMGAS memory system behavior
while maintaining backward compatibility and improving overall code quality.

dingyi222666 committed:
feat(long-memory): improve extraction robustness and configuration structure

- Restructure configuration to move longMemoryExtractInterval from HippoRAG to basic settings
- Enhance YAML parsing robustness with fallback mechanisms for memory extraction
- Update memory extraction prompt to use YAML format with comprehensive documentation
- Improve EMGAS extractor array handling with safer concat operations
- Add retrievalLayer field to EnhancedMemory for better memory tracking
- Update tool descriptions to reflect new USER layer as default instead of PRESET_USER
- Standardize localization file structure with proper indentation and hierarchy
- Fix vector store initialization checks in HippoRAG layer
- Update documentToEnhancedMemory calls to include layer info parameter
- Remove deprecated configuration parameters and consolidate extraction settings

This refactor improves extraction reliability, standardizes configuration structure,
and provides better error handling throughout the memory system.

dingyi222666 committed:
feat(long-memory): implement Basic memory layer and improve prompt formatting

- Add BasicMemoryLayer implementation with direct database storage and retrieval
- Implement complete CRUD operations (add, retrieve, delete, clear memories)
- Add automatic expired memory cleanup with periodic intervals
- Update layer engine configuration to include Basic layer support
- Improve prompt formatting with structured XML-like tags for better context parsing
- Enhance document metadata rendering with structured format and proper spacing
- Update default layer engine configuration with multi-layer setup
- Remove debug console.log statements and clean up service initialization
- Add proper Basic layer integration in init_layer.ts

The Basic layer provides simple memory storage without vector search or graph processing,
making it suitable for lightweight deployments and fallback scenarios.

dingyi222666 committed:
fix(vector-store-service): refactor prompt template management and fix JSON formatting

- Refactor PROMPT_LIST to use centralized management in PromptManager
- Move prompt template imports into registerAllPrompts function for better initialization control
- Add explicit prompt registration call in main index.ts to ensure proper initialization
- Fix JSON formatting in ner_query template output with proper escaping
- Improve code organization by separating prompt registration from module imports

This refactor resolves prompt template initialization issues and ensures proper
JSON output formatting for named entity recognition queries.

dingyi222666 committed:
style(long-memory): fix code formatting in BasicMemoryLayer

- Fix multi-line string formatting for better readability
- Add proper newline at end of file
- Improve code style consistency with project standards

Minor formatting improvements to maintain code quality standards.

dingyi222666 commented on packages/long-memory/src/layers/hippo/layer.ts:
```diff
@@ -0,0 +1,533 @@
+            const docs = filtered.slice(0, Math.max(1, topK)).map((s) => s.doc)
+
+            if (docs.length > 0) {
+                const nowISO = new Date().toISOString()
+                for (const d of docs) {
+                    const meta = d.metadata ?? {}
+                    meta.last_accessed = nowISO
+                    meta.access_count = Number(meta.access_count ?? 0) + 1
+                    d.metadata = meta
+                    // keep cache in sync
+                    const key: string =
+                        (meta.simhash as string) ||
+                        computeSimHashHex(d.pageContent)
+                    this.simhashDocCache.set(key, d)
+                }
+
+                for (const d of docs) {
+                    await this.vectorStore.editDocument(d.id, d)
+                }
+
+                try {
+                    await this.vectorStore.save()
+                } catch (e) {
+                    logger?.debug('save after access update failed', e)
+                }
+            }
```
无需，默认就支持，这是使用 chatluna 的

dingyi222666 committed:
fix(long-memory,core,vector-store-service): improve robustness and error handling

- Add HTML escaping for metadata and document IDs in prompt formatting to prevent XML injection
- Improve model initialization checks with proper null validation in chat middleware
- Fix BasicMemoryLayer to use memoryId directly instead of computed hash for simpler storage
- Add memory ID validation in BasicMemoryLayer initialization
- Enhance error handling in HippoRAG retrieval with try-catch for vector store operations
- Add stable ID generation for EMGAS memories to prevent overwrites
- Move prompt registration to async context on ready event for proper initialization timing
- Convert registerAllPrompts to async function with Promise.all for parallel template loading

These improvements enhance system stability, prevent data corruption, and ensure
proper initialization order across all memory layer implementations.

dingyi222666 committed:
fix(long-memory): correct localization schema structure and formatting

- Fix layerEngines schema structure by changing 'engines' to 'engine' for proper field mapping
- Remove trailing periods from layer type descriptions for consistency
- Standardize formatting in both English and Chinese localization files
- Ensure schema field names match the actual configuration structure

This fixes configuration UI display issues where layer engines were not
properly recognized due to schema field name mismatch.

dingyi222666 committed:
refactor(long-memory): improve security and clean up imports

- Remove unused resolveLongMemoryId import from basic layer
- Add input validation and path sanitization in emgas layer
- Prevent directory traversal vulnerabilities in getGraphFilePath
- Ensure memoryId is properly validated before file operations

dingyi222666 committed:
feat(long-memory): add memory_update tool for updating existing memories

- Add memory_update tool registration in plugin system
- Implement MemoryUpdateTool class with delete-then-add strategy
- Support updating memories across user, preset, and global layers
- Include comprehensive schema validation and error handling
- Provide detailed usage guidelines in tool description

dingyi222666 committed:
docs(long-memory): update EMGAS parameter descriptions in en-US and zh-CN locales

Enhanced documentation for EMGAS long-term memory settings including:
- emgasDecayRate: Renamed to Temporal decay Lambda (λ) with detailed explanation
- emgasPruneThreshold: Clarified pruning mechanism for low-activation nodes
- emgasFiringThreshold: Renamed to Firing threshold (F) with expanded description
- emgasPropagationDecay: Renamed to Propagation decay (D) with improved clarity
- emgasMaxIterations: Detailed impact on retrieval latency and association depth
- emgasTopN: Updated description for candidate memory retrieval control

Updated both English and Chinese locale files to ensure consistency and clarity
of advanced memory graph configuration parameters.

dingyi222666 commented on packages/long-memory/src/layers/basic/layer.ts:
```diff
@@ -0,0 +1,168 @@
+export class BasicMemoryLayer<
+    T extends MemoryRetrievalLayerType = MemoryRetrievalLayerType
+> extends BaseMemoryRetrievalLayer<T> {
+    private docstore: DataBaseDocstore
```
无需

dingyi222666 committed:
fix(long-memory): use correct layer type in edit memory command

- Replace hardcoded MemoryRetrievalLayerType.GLOBAL with parsedLayerType
- Ensure memory retrieval uses the user-specified layer type instead of always global
- Fixes layer type mismatch in memory editing functionality

dingyi222666 commented:
等我起床后 27 号晚上大概率合并，差不多了

dingyi222666 commented:
今天没睡好头晕了，明天 28 号合

dingyi222666 committed:
feat(long-memory,core): add initMemoryLayers to all tools and improve layer initialization

- Add initMemoryLayers calls to MemorySearchTool, MemoryDeleteTool, and MemoryUpdateTool for consistent layer initialization
- Fix database injection for DataBaseDocstore to handle optional database dependency
- Improve memory layer initialization logic to handle partial layer types
- Replace logging levels from info to debug for better log management
- Remove deprecated parse utility functions and update chat history extraction
- Update default enabled layers configuration to Global and User
- Clean up import statements and remove unused code

This ensures all memory tools properly initialize their layers before operations,
improving reliability and preventing initialization-related errors.

dingyi222666 commented on packages/long-memory/src/plugins/tool.ts:
```diff
@@ -77,15 +87,27 @@ export class MemorySearchTool extends StructuredTool {
-                config.configurable.conversationId,
-                input.content,
+            const parsedLayerType =
                 input.layer != null
                     ? input.layer.map(
                           (layer) =>
                               MemoryRetrievalLayerType[layer.toUpperCase()]
                       )
-                    : MemoryRetrievalLayerType.PRESET_USER
+                    : MemoryRetrievalLayerType.USER
+
+            await this.ctx.chatluna_long_memory.initMemoryLayers(
+                config.configurable.conversationId,
+                {
+                    presetId: config.configurable.preset,
+                    userId: config.configurable.userId
+                },
+                parsedLayerType
+            )
+
+            const result = await this.ctx.chatluna_long_memory.retrieveMemory(
+                config.configurable.conversationId,
+                input.content,
+                parsedLayerType
             )
```
no

dingyi222666 commented on packages/long-memory/src/plugins/tool.ts:
```diff
@@ -259,12 +302,132 @@ export class MemoryDeleteTool extends StructuredTool {
+                layers
+            )
+
+            return `Successfully updated ${input.memoryIds.length} memories with ${enhancedMemories.length} new memories.`
```
no

dingyi222666 committed:
feat(long-memory): implement atomic memory updates and secure logging

Add atomic updateMemories API to prevent data loss during memory updates:
- Implement service-level updateMemories with ID preservation and rollback capability
- Add getMemoriesByIds method for backup support before updates
- Update MemoryUpdateTool to use atomic API instead of separate delete+add operations
- Add validation to ensure memoryIds count matches newMemories count
- Implement rollback mechanism to restore original memories on failure

Add secure logging to prevent sensitive data leakage:
- Create configurable secure logging utility with content truncation and fingerprinting
- Add CHATLUNA_SAFE_LOG_LENGTH environment variable for safe content length (default: 150)
- Replace raw content logging with safe preview and SHA-256 fingerprint
- Apply secure logging to model output in chat history extraction

This ensures memory update operations are atomic and prevents data loss while
protecting sensitive information from appearing in debug logs.
